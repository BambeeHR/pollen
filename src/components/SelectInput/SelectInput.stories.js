import BaseIcon from '../BaseIcon/BaseIcon.vue';
import SelectInput from './SelectInput.vue';

export default {
  title: 'SelectInput',
  component: SelectInput,
  parameters: {
    componentSubtitle: 'A select input capable of selecting multiple options.',
  },
};

const options = [
  'Option 1',
  {
    label: 'Option 2',
    value: 'foo',
  },
  {
    label: 'Option 3',
    value: 'bar',
  },
  {
    label: 'Test',
    value: 'test',
  },
  'Something',
  {
    label: 'Hello',
    value: 'World',
  },
];

export const Default = () => ({
  components: { SelectInput },
  data() {
    return {
      options,
      value: null,
    };
  },
  template: `
    <SelectInput v-model="value" :options="options" />
  `,
});

export const MultiSelect = () => ({
  components: { SelectInput },
  data() {
    return {
      options,
      value: null,
    };
  },
  template: `
    <SelectInput v-model="value" :options="options" multiple />
  `,
});

/* eslint-disable no-param-reassign */
[Default, MultiSelect].forEach((item) => {
  item.story = {
    parameters: {
      jest: ['SelectInput.test.js'],
    },
  };
});
