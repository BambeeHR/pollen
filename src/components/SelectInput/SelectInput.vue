<template>
  <TextInput
    :id="id"
    v-model="filter"
    class="select-input"
    v-bind="passDownProps"
    @input="handleInput"
    @focus="isDropdownActive = true"
  >
    <template v-if="multiple" #pre-field>selected options go here</template>
    <template #post-field>
      <transition>
        <ul>
          <li v-for="item in mappedOptions" :key="item.value">
            <button type="button" @click="handleSelect(value)">
              {{ label }}
            </button>
          </li>
        </ul>
      </transition>
    </template>
  </TextInput>
</template>

<script>
import TextInput from '../TextInput/TextInput.vue';
import mapOptions from '../../util/mapOptions';
import { Icons } from '../BaseIcon/BaseIcon.vue';

export default {
  components: { TextInput },
  inheritAttrs: false,
  props: {
    /** An intial filter value to set. */
    initialFilter: {
      type: String,
      default: '',
    },
    /** If true, allows for selecting multiple options. */
    multiple: {
      type: Boolean,
      default: false,
    },
    /**
     * The available options for this component. This is an array of values.
     * Each individual value can be either a string, or an object with a `label`
     * and `value` key. The `label` is what the user sees. The `value` is what
     * will be emitted as an `input` event.
     */
    options: {
      type: Array,
      default: () => [],
    },
    /**
     * The selected value. Compatible with v-model. If `multiple` is true, this
     * must be an array of string values.
     */
    value: {
      type: [String, Array],
      default: '',
    },
  },
  data() {
    return {
      filter: this.initialFilter,
      isDropdownActive: false,
    };
  },
  computed: {
    mappedOptions() {
      const filterString = this.filter.toLowerCase().trim();
      const isFilterStringInLabel = (item) =>
        item.label.toLowerCase().includes(filterString);
      return mapOptions(this.options).filter(isFilterStringInLabel);
    },
    passDownProps() {
      return {
        ...this.$attrs,
        postIcon: this.isDropdownActive ? Icons.CARET_UP : Icons.CARET_DOWN,
      };
    },
  },
};
</script>
