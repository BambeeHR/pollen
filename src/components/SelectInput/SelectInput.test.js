import { mount } from '@vue/test-utils';
import SelectInput from './SelectInput.vue';

const options = [
  'Option 1',
  {
    label: 'Option 2',
    value: 'foo',
  },
  {
    label: 'Option 3',
    value: 'bar',
  },
  {
    label: 'Test',
    value: 'test',
  },
  'Something',
  {
    label: 'Hello',
    value: 'World',
  },
];

describe('SelectInput', () => {
  test('renders correctly', () => {
    const wrapper = mount(SelectInput, {
      propsData: {
        options,
        value: 'test',
        id: 'test',
        label: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  // TODO(jjandoc): Not sure of the best way to test this since it's using the
  // third-party vue-select component, and there seems to be some issues with
  // properly mounting the component for tests. Skipping for now.
  test.skip('emits an input event', async () => {
    const pressDown = new KeyboardEvent('keydown', { keyCode: 37 });
    const pressEnter = new KeyboardEvent('keydown', { key: 'Enter' });
    const Component = {
      components: { SelectInput },
      template: `<SelectInput v-model="value" :options="options" />`,
      data() {
        return {
          options,
          value: '',
        };
      },
    };
    const wrapper = mount(Component);
    const input = wrapper.find('input');
    await input.trigger('focus');
    document.dispatchEvent(pressDown);
    document.dispatchEvent(pressEnter);
    expect(wrapper.vm.value).toEqual('foo');
  });
});
