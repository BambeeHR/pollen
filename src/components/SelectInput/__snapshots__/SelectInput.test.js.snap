// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput renders correctly 1`] = `
<label
  class="form-field select-input select-input--standard select-input--normal"
  for="test"
>
  <div
    class="field-label field-label--standard field-label--normal"
  >
    
    test
  
  </div>
   
  <div
    class="field-container field-container--standard field-container--normal"
  >
    <div
      class="v-select vs--single vs--searchable"
      dir="auto"
    >
       
      <div
        aria-expanded="false"
        aria-label="Search for option"
        aria-owns="vs1__listbox"
        class="vs__dropdown-toggle"
        id="vs1__combobox"
        role="combobox"
      >
        <div
          class="vs__selected-options"
        >
          <div
            class="select-input__selected"
          >
            
        Test
      
          </div>
           
          <input
            aria-autocomplete="list"
            aria-controls="vs1__listbox"
            aria-labelledby="vs1__combobox"
            autocomplete="off"
            class="vs__search"
            type="search"
          />
        </div>
         
        <div
          class="vs__actions"
        >
          <button
            aria-label="Clear Selected"
            class="vs__clear"
            title="Clear Selected"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="base-icon svg-inline--fa fa-times fa-w-10"
              data-icon="times"
              data-prefix="fal"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class=""
                d="M193.94 256L296.5 153.44l21.15-21.15c3.12-3.12 3.12-8.19 0-11.31l-22.63-22.63c-3.12-3.12-8.19-3.12-11.31 0L160 222.06 36.29 98.34c-3.12-3.12-8.19-3.12-11.31 0L2.34 120.97c-3.12 3.12-3.12 8.19 0 11.31L126.06 256 2.34 379.71c-3.12 3.12-3.12 8.19 0 11.31l22.63 22.63c3.12 3.12 8.19 3.12 11.31 0L160 289.94 262.56 392.5l21.15 21.15c3.12 3.12 8.19 3.12 11.31 0l22.63-22.63c3.12-3.12 3.12-8.19 0-11.31L193.94 256z"
                fill="currentColor"
              />
            </svg>
          </button>
           
          <svg
            class="base-icon base-icon--custom vs__open-indicator"
            role="presentation"
          />
           
          <div
            class="vs__spinner"
            style="display: none;"
          >
            Loading...
          </div>
        </div>
      </div>
       
      <transition-stub
        name="vs__fade"
      >
        <ul
          id="vs1__listbox"
          role="listbox"
          style="display: none; visibility: hidden;"
        />
      </transition-stub>
       
    </div>
  </div>
   
  <!---->
</label>
`;
